cc/main.cc

```diff
@@ -464,46 +473,244 @@ class SelfPlayer {
   uint64_t flags_timestamp_ = 0;
 };
 
+class Evaluator {
+  // A barrier that blocks threads until the number of waiting threads reaches
+  // the 'count' threshold. Allows decrementing the threshold to handle the tail
+  // of a work queue where some threads exit early.
+  class Barrier {
+   public:
+    explicit Barrier(size_t count)
+        : count_(count), num_waiting_(0), generation_(0) {}
+
+    void Wait() {
+      absl::MutexLock lock(&mutex_);
+      if (++num_waiting_ == count_) {
+        IncrementGeneration();
+      } else {
+        auto generation = generation_;
+        while (generation != generation_) {
+          cond_var_.Wait(&mutex_);
+        }
+      }
+    }
+
+    void DecrementCount() {
+      absl::MutexLock lock(&mutex_);
+      if (num_waiting_ == --count_) {
+        IncrementGeneration();
+      }
+    }
+
+   private:
+    void IncrementGeneration() EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+      ++generation_;
+      num_waiting_ = 0;
+      cond_var_.SignalAll();
+    }
+
+    absl::Mutex mutex_;
+    absl::CondVar cond_var_;
+    size_t count_ GUARDED_BY(&mutex_);
+    size_t num_waiting_ GUARDED_BY(&mutex_);
+    size_t generation_ GUARDED_BY(&mutex_);
+  };
+
+  // References a pointer to an actual DualNet. Allows updating the pointer
+  // after the MctsPlayer has been constructed.
+  class WrappedDualNet : public DualNet {
+   public:
+    explicit WrappedDualNet(const std::unique_ptr<DualNet>& dual_net)
+        : dual_net_(dual_net) {}
+
+   private:
+    void RunMany(std::vector<const BoardFeatures*> features,
+                 std::vector<Output*> outputs, std::string* model) override {
+      dual_net_->RunMany(std::move(features), std::move(outputs), model);
+    };
+
+    const std::unique_ptr<DualNet>& dual_net_;
+  };
+
+  struct Model {
+    std::string name;
+    std::unique_ptr<DualNetFactory> factory;
+    std::atomic<int> black_wins;
+    std::atomic<int> white_wins;
+  };
+
+ public:
+  void Run() {
+    auto start_time = absl::Now();
+
+    auto create_model = [](const std::string& model_path) {
```

Comment:
Why isn't this a constructor of Model?

Reply:
It's not a constructor because I wanted to keep it a struct, but maybe I'm interpreting the cpp-style too strictly that structs should not contain any functionality.

Reply:
Yeah, I can understand that. I have no objection to structs having a constructor.